### PR TITLE
feature/plugin-170-auth-update

### PR DESCRIPTION
**Pull request by ardian4**
Added: Ability to provide Service Account type as JSON

Jira Task: https://issues.cask.co/browse/PLUGIN-170

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java [hunk 1]
```diff
@@ -16,8 +16,12 @@
  */
 public class GCPConfig extends PluginConfig {
   public static final String NAME_PROJECT = "project";
+  public static final String NAME_SERVICE_ACCOUNT_TYPE = "serviceAccountType";
   public static final String NAME_SERVICE_ACCOUNT_FILE_PATH = "serviceFilePath";
+  public static final String NAME_SERVICE_ACCOUNT_JSON = "serviceJson";
   public static final String AUTO_DETECT = "auto-detect";
+  public static final String SERVICE_ACCOUNT_TYPE_FILE_PATH = "filePath";
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java]: Typo: `SERVICE_ACCOUNT_FILE_PATH`

File: src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java [hunk 2]
```diff
@@ -16,8 +16,12 @@
...
+  public static final String SERVICE_ACCOUNT_TYPE_JSON = "json";
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java]: Typo: `SERVICE_ACCOUNT_JSON`

File: src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java [hunk 3]
```diff
@@ -26,6 +30,12 @@
   @Nullable
   protected String project;
 
+  @Name(NAME_SERVICE_ACCOUNT_TYPE)
+  @Description("Service account which can be either file path or JSON content.")
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java]: This description does not match the service account type.

File: src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java [hunk 4]
```diff
@@ -34,6 +44,12 @@
   @Nullable
   protected String serviceFilePath;
 
+  @Name(NAME_SERVICE_ACCOUNT_JSON)
+  @Description("Content of the service account file.")
+  @Macro
+  @Nullable
+  protected String serviceJson;
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java]: `serviceAccountJson`

File: src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java [hunk 5]
```diff
@@ -64,6 +80,40 @@ public String getServiceAccountFilePath() {
...
+
+  public String getServiceAccountType() {
+    if (containsMacro(NAME_SERVICE_ACCOUNT_TYPE) || Strings.isNullOrEmpty(serviceAccountType)) {
+      return null;
+    }
+    return serviceAccountType;
+  }
+
+  public Boolean isServiceAccountJson() {
+    return getServiceAccountType() == null ?
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java]: we should call the method just once and have a local variable for it.

File: src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java [hunk 6]
```diff
@@ -64,6 +80,40 @@ public String getServiceAccountFilePath() {
...
+      null : getServiceAccountType().equals(SERVICE_ACCOUNT_TYPE_JSON);
+  }
+
+  public Boolean isServiceAccountFilePath() {
+    return getServiceAccountType() == null ?
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java]: should not call same method multiple times

File: widgets/GCSArgumentSetter-action.json [hunk 7]
```diff
@@ -82,12 +82,12 @@
     {
       "name": "ServiceAuthenticationTypeJSON",
       "condition": {
-        "expression": "serviceAccountType == 'JSON'"
+        "expression": "serviceAccountType == 'json'"
```
Comment (CuriousVini) [inline widgets/GCSArgumentSetter-action.json]: why is this change needed?

File: widgets/GCSArgumentSetter-action.json [hunk 8]
```diff
@@ -33,35 +33,35 @@
...
           "widget-attributes": {
             "default": "auto-detect"
           }
         },
         {
           "widget-type": "textbox",
-          "label": "Service Account JSON",
-          "name": "serviceAccountJSON"
+          "label": "JSON",
+          "name": "serviceJson"
```
Comment (CuriousVini) [inline widgets/GCSArgumentSetter-action.json]: why are we changing this to serviceJson?

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java [hunk 9]
```diff
@@ -49,11 +64,55 @@ public static ServiceAccountCredentials loadServiceAccountCredentials(String pat
...
+    throws IOException {
+    if (isServiceAccountFilePath) {
+      return loadServiceAccountCredentials(content);
+    }
+    InputStream jsonInputStream = new ByteArrayInputStream(content.getBytes());
+    return ServiceAccountCredentials.fromStream(jsonInputStream);
+  }
+
+  public static String extractPrivateKey(ServiceAccountCredentials credentials) {
+    return String.format(PRIVATE_KEY_WRAP,
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:78]: why are we having this private key wrap? Wouldnt it always have this wrap in the service account key? We should not modify anything in private key

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java [hunk 10]
```diff
@@ -49,11 +64,55 @@ public static ServiceAccountCredentials loadServiceAccountCredentials(String pat
...
+  }
+
   public static Map<String, String> getFileSystemProperties(GCPConfig config, String path,
                                                             Map<String, String> properties) {
-    String serviceAccountFilePath = config.getServiceAccountFilePath();
-    if (serviceAccountFilePath != null) {
-      properties.put(CLOUD_JSON_KEYFILE, serviceAccountFilePath);
+    try {
+      properties.putAll(generateAuthProperties(config.getServiceAccount(), config.isServiceAccountFilePath(),
+                                               CLOUD_JSON_KEYFILE_PREFIX));
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:132]: why do we need to pass this? There is no need of this prefix.

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java [hunk 11]
```diff
@@ -49,11 +64,55 @@ public static ServiceAccountCredentials loadServiceAccountCredentials(String pat
...
+    return ServiceAccountCredentials.fromStream(jsonInputStream);
+  }
+
+  public static String extractPrivateKey(ServiceAccountCredentials credentials) {
+    return String.format(PRIVATE_KEY_WRAP,
+                         Base64.getEncoder().encodeToString(credentials.getPrivateKey().getEncoded()));
+  }
+
+  public static Map<String, String> generateAuthProperties(String serviceAccount,
+                                                           boolean isServiceAccountFilePath,
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java]: we should just pass the type here

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java [hunk 12]
```diff
@@ -49,11 +64,55 @@ public static ServiceAccountCredentials loadServiceAccountCredentials(String pat
...
+    InputStream jsonInputStream = new ByteArrayInputStream(content.getBytes());
+    return ServiceAccountCredentials.fromStream(jsonInputStream);
+  }
+
+  public static String extractPrivateKey(ServiceAccountCredentials credentials) {
+    return String.format(PRIVATE_KEY_WRAP,
+                         Base64.getEncoder().encodeToString(credentials.getPrivateKey().getEncoded()));
+  }
+
+  public static Map<String, String> generateAuthProperties(String serviceAccount,
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:100]: Add javadoc on what properties are being set in this method

File: widgets/GCSArgumentSetter-action.json (hunk 7 above)
Comment (ardian4) [reply widgets/GCSArgumentSetter-action.json]: GCSArgumentSetter already had this feature and was merged to develop, since those code changes will be used by all other GC plugins we had to unify GCSArgumentSetter.

File: widgets/GCSArgumentSetter-action.json (hunk 8 above)
Comment (ardian4) [reply widgets/GCSArgumentSetter-action.json]: GCSArgumentSetter already had this feature and was merged to develop, since those code changes will be used by all other GC plugins we had to unify GCSArgumentSetter.

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java (hunk 9 above)
Comment (ardian4) [reply src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:78]: ``ServiceAccountCredentials.getPrivateKey().getEncoded()`` will not give the private key with prefix ``-----BEGIN PRIVATE KEY-----`` xxxxx ``----- END PRIVATE KEY-----``, is there any other way how get the key from ServiceAccountCredentials?

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java (hunk 10 above)
Comment (ardian4) [reply src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:132]: This is used in following places:

https://github.com/data-integrations/google-cloud/blob/ff04f4783a3bd82dafb9d87bff150f6061bab322/src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java#L139


https://github.com/data-integrations/google-cloud/blob/48a86cecc1cdf7e45b95bfbe76b19662754b8f2a/src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java#L155-L158

File: widgets/GCSArgumentSetter-action.json (hunk 7 above)
Comment (CuriousVini) [reply widgets/GCSArgumentSetter-action.json]: But why are we changing service account type from json -> JSON?

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java [hunk 13]
```diff
@@ -49,11 +64,55 @@ public static ServiceAccountCredentials loadServiceAccountCredentials(String pat
...
+      if (isServiceAccountFilePath) {
+        properties.put(String.format("%s.%s", prefix, CLOUD_JSON_KEYFILE_SUFFIX), serviceAccount);
+        continue;
+      }
+      ServiceAccountCredentials credentials = loadServiceAccountCredentials(serviceAccount, false);
+
+      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_EMAIL_SUFFIX), credentials.getClientEmail());
+      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_PRIVATE_KEY_ID_SUFFIX),
+                     credentials.getPrivateKeyId());
+      if (privateKeyData == null) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:119]: Isnt private key always null? Why do we have this check?

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java (hunk 9 above)
Comment (CuriousVini) [reply src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:78]: I see

File: widgets/GCSArgumentSetter-action.json (hunk 8 above)
Comment (CuriousVini) [reply widgets/GCSArgumentSetter-action.json]: Why are we changing name? I think it should be `serviceAccountJSON`

File: widgets/GCSArgumentSetter-action.json [hunk 14]
```diff
@@ -82,12 +82,12 @@
...
       "name": "ServiceAuthenticationTypeJSON",
       "condition": {
-        "expression": "serviceAccountType == 'JSON'"
+        "expression": "serviceAccountType == 'json'"
       },
       "show": [
         {
           "type": "property",
-          "name": "serviceAccountJSON"
+          "name": "serviceJson"
```
Comment (CuriousVini) [inline widgets/GCSArgumentSetter-action.json]: Shouldnt this be `serviceAccountJSON`?

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java (hunk 13 above)
Comment (ardian4) [reply src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:119]: Is kind of optimization. Check is inside loop, and private key will be always same, in first pass we do extract and save in  local variable, so in the second pass key will be already extracted. We can remove the check ?

File: widgets/GCSArgumentSetter-action.json [hunk 15]
```diff
@@ -33,34 +33,34 @@
...
+                "id": "JSON",
+                "label": "JSON"
               }
             ]
           }
         },
         {
           "widget-type": "textbox",
-          "label": "Service Account File Path",
+          "label": "File Path",
```
Comment (CuriousVini) [inline widgets/GCSArgumentSetter-action.json]: Lets keep it same `Service Account File Path`

File: widgets/GCSArgumentSetter-action.json [hunk 16]
```diff
@@ -33,34 +33,34 @@
...
           "name": "serviceFilePath",
           "widget-attributes": {
             "default": "auto-detect"
           }
         },
         {
           "widget-type": "textbox",
-          "label": "Service Account JSON",
+          "label": "JSON",
```
Comment (CuriousVini) [inline widgets/GCSArgumentSetter-action.json]: Lets keep it same `Service Account JSON`

**Review by CuriousVini — APPROVED** [review]
2 minor comments. lgtm 👍 
Please squash the commits.

**Comment by ardian4** [comment]
> 2 minor comments. lgtm 👍
> Please squash the commits.

@CuriousVini squashed.
